sections/producer.tex

```diff
@@ -1,5 +1,26 @@
+
+The  \CCSPStlye{Producer} class \code{mc\_weight} reads in the \code{genWeight} column and, if existent, the \code{LHEWeight} column, both stored in \code{events}. Since these columns are required, they are both added to the \code{uses} set of \code{mc\_weight}. By extension, when we call \code{mc\_weight} in our \code{uses} set, we are calling these columns as well. The \code{mc\_weight} class simply decides which one of these weights to use and saves the decision as a new column, also named \code{mc\_weight}, which is included in its \code{produces} set. Again, by extension, we must add the \code{mc\_weight} class to our own \code{produces} set such that this new column also gets created and saved to disk. 
+
+Meanwhile, the \code{category\_ids} class assigns each event an array of category ids, which it stores as a new column also named \code{category\_ids}. Thus, we must also add this class to our \code{produces} set. The topic of defining categories is discussed in detail in the Section \ref{sec:categories}. Now that we have access to both these \CCSPStlye{Producer} classes, the \code{cutflow\_features} class can use them to attribute MC weights (if the dataset passed to it is tagged as an MC dataset) and category ids to \code{events}. It then creates a new column in the updated \code{events} object named \code{cutflow.jet1\_pt} which saves the transverse momentum of the most energetic jet in each event stored in \code{Jet.pt\text{[:,0]}}. If the event does not contain jets, it instead saves an \code{EMPTY\_FLOAT} value. 
+
+The last \CCSPStlye{Producer} class defined is \code{example} and follows the same structure as the  two previously explained \CCSPStlye{Producer}s. First, it starts by creating the \code{cutflow.jet1\_pt} column by using the \CCSPStlye{Producer} class \code{features} called at \code{\text{events=self[features](events, **kwargs)}}. It then applies category ids and determinist seeds to the updated \code{events} object. Lastly, it applies normalization weights and muon weights only to \code{events} tagged as originating from a MC dataset. 
+
+\begin{exercise}{Understanding some basic Producers}
+	Familiarize yourself with the \CCSPStlye{Producer} classes mentioned above.
+\end{exercise}
+
+In this \code{H4L} analysis we want to calculate the four-lepton invariant mass, so that we can see the resonant Higgs peak. Note that to perform four-vector calculations, you need to import \code{attach\_coffea\_behavior} from \code{columnflow.production.util}. You will need to use kinematic information from both the \code{events.Electron} and \code{events.Muon} collections and create a new column which stores your calculated invariant mass.
+
+\begin{exercise}{Writing a Producer}
```

Comment:
Suggest to include path to solution here

```suggestion
\begin{exercise}{Writing a Producer}{h4l/production/invariant_mass.py}
```

Reply:
I included it. Thanks!